New logic for filters

Denis-Shtanskiy opened the pull request:
Добавил фильтры, для всех полей которые будут использоваться на фронте. Обновил вьюсет, для конечной фильтрации кверисета, чтобы применялся фильтр уже после имеющейся выдачи.

Denis-Shtanskiy committed:
add filters, update viewset

ArtemKAF reviewed (commented):
https://github.com/Pet-projects-CodePET/Backend/blob/59f499ff2d4dcbc0c5dae133dcbc6c7b72ba6e11/infra/docker-compose-prod.yaml#L21-L21
Тут POSTGRES_DB нужна переменная, а не POSTGRES_NAME. Так как флаг -d указывает какую БД проверять на доступность.

https://github.com/Pet-projects-CodePET/Backend/blob/59f499ff2d4dcbc0c5dae133dcbc6c7b72ba6e11/src/backend/api/v1/projects/views.py#L91-L93
Здесь излишне снова вызывать метод filter_queryset, так как он будет вызван при вызове родительской реализации метода get_queryset. А так получается двойная работа.

https://github.com/Pet-projects-CodePET/Backend/blob/59f499ff2d4dcbc0c5dae133dcbc6c7b72ba6e11/src/backend/api/v1/projects/filters.py#L13-L13
Тут излишне lookup_expr.

https://github.com/Pet-projects-CodePET/Backend/blob/59f499ff2d4dcbc0c5dae133dcbc6c7b72ba6e11/src/backend/api/v1/projects/filters.py#L12-L12
Тут, вроде бы, решили реализовать, как и с полем ended границу дат задавать.

https://github.com/Pet-projects-CodePET/Backend/blob/59f499ff2d4dcbc0c5dae133dcbc6c7b72ba6e11/src/backend/api/v1/projects/filters.py#L14-L19
Это вычисляемое поле. Оно не хранится в БД, а фильтрсету нужно именно по полям из БД реализовывать фильтрацию. Поэтому такая реализация вызывает ошибку
```
django.core.exceptions.FieldError: Cannot resolve keyword 'recruitment_status' into field. Choices are: busyness, created, creator, creator_id, description, directions, email, ended, id, link, modified, name, owner, owner_id, participants, phone_number, project_specialists, started, status, telegram_nick
```
https://github.com/Pet-projects-CodePET/Backend/blob/59f499ff2d4dcbc0c5dae133dcbc6c7b72ba6e11/src/backend/api/v1/projects/filters.py#L20-L23
и
https://github.com/Pet-projects-CodePET/Backend/blob/59f499ff2d4dcbc0c5dae133dcbc6c7b72ba6e11/src/backend/api/v1/projects/filters.py#L26-L26

Пока, судя по дизайну в figma, нет необходимости требовать от фронта именно строки для фильтрации. Фронты легко смогут идентификаторы отправлять вместо названий для объектов. Это избавит нас от необходимости в БД индексы создавать для этих полей.
Другое дело, если фронт предоставит пользователям вводить "от руки", тогда останется только так, как у тебя реализовывать.

Denis-Shtanskiy committed:
add filters for specialists, fix comments, add feature my-projects

Denis-Shtanskiy commented:
Добавил ещё фильтры на будущее для специалистов. Изменил пре-коммит, так как устал что у меня версия 3.12.2 и он не работал, изменил на просто 3.12, добавил пока что в предполагаемом варианте фичу по превью в личном кабинете

Denis-Shtanskiy committed:
fix filter, delete comments

ArtemKAF reviewed (changes requested):
https://github.com/Pet-projects-CodePET/Backend/blob/9377ebcc3c9c9f5959484baa5e0db1fee4109b90/src/backend/api/v1/profile/views.py#L12-L12
и
https://github.com/Pet-projects-CodePET/Backend/blob/9377ebcc3c9c9f5959484baa5e0db1fee4109b90/src/backend/api/v1/profile/views.py#L33-L33
После изменения имени класса фильтрации нужно и здесь его изменить.


https://github.com/Pet-projects-CodePET/Backend/blob/9377ebcc3c9c9f5959484baa5e0db1fee4109b90/src/backend/api/v1/profile/views.py#L35-L46
Здесь мжно избавиться от нескольких строк кода.
Ветка else абсолютно не обязательна и можно вместо нее сразу возвращать queryset отфильтрованный по visible_status=Profile.ALL


https://github.com/Pet-projects-CodePET/Backend/blob/9377ebcc3c9c9f5959484baa5e0db1fee4109b90/src/backend/api/v1/projects/views.py#L88-L93
Это дублирование кода. Этот метод вьюсет проектов унаследует от BaseProjectViewSet


https://github.com/Pet-projects-CodePET/Backend/blob/9377ebcc3c9c9f5959484baa5e0db1fee4109b90/src/backend/api/v1/projects/filters.py#L15-L21
Здесь излишне использовать MultipleChoice и похоже, что без добавления аннотированного поля в кверисет нам не обойтись. Твоя реализация не работает и, скорее всего не заработает, пока это поле вычисляется на лету внутри сериализатора. Думаю, что придется добавить аннотированное поле recruitment_status в queryset, чтобы можно было по нему фильтровать.
https://github.com/Pet-projects-CodePET/Backend/blob/9377ebcc3c9c9f5959484baa5e0db1fee4109b90/src/backend/api/v1/projects/filters.py#L44-L55
Этот метод ничего не фильтрует, как минимум потому, что в value список попадает при использовании MultipleChoice.
Плюс есть опечатка в is_required.

https://github.com/Pet-projects-CodePET/Backend/blob/9377ebcc3c9c9f5959484baa5e0db1fee4109b90/src/backend/api/v1/projects/filters.py#L26-L28
В direction не работает, потому, что нужно указать поле directions, а в level не работает, потому, что поле, по которому нужно фильтровать находится в специалистах, связанных с проектом.

P.S.: Поправочка. Удалось получить результаты похожие на правду для фильтрации по recruitment_status применив следующие фильтрации:
```python
if value[0] == "1":
    queryset = queryset.exclude(
        Q(project_specialists__is_required=False)
        | Q(project_specialists=None)
    )
    return queryset.filter(
        ~Q(project_specialists__is_required=True)
        | Q(project_specialists=None)
    )
```
P.P.S.: Ты был изначально прав. Всетаки нужен именно SpecialistsFilterSet. Только не привязывай его сейчас к вьюсетам. Те, которые сейчас в коде их нужно переделывать довольно сильно. Адриан переделает и привяжет или привяжешь сам, когда вьюсет именно для специалистов появится.

Denis-Shtanskiy committed:
Merge branch 'develop' into feature/filters

Denis-Shtanskiy committed:
Merge remote-tracking branch 'origin/develop' into feature/filters

Denis-Shtanskiy committed:
fix: Добавил изменения по замечаниям
- ТАк же убрал сборку и пуш из CI
- Добавил МОсковское время в настройки
- Проверил работоспособность моего кода, прошлого, все работало, при выборе 0 и 1 через NumberFilter, так более читаемо, как предложил Артем

Denis-Shtanskiy commented:
Так же убрал сборку и пуш из CI, при мердже в девелоп пофиксит.
 Добавил Московское время в настройки
 Проверил работоспособность моего кода, прошлого, все работало, при выборе 0 и 1 через NumberFilter, так более читаемо, как предложил Артем.
 Идея подставлять миксин - не сработала, тогда пришлось бы явно из вью передавать в поля экземпляр, выглядело как улсожнение во вью, но упрощение в фильтре.

Denis-Shtanskiy committed:
fix: пропустил файлы для CI

ArtemKAF commented on src/backend/manage.py:
Верни, пожалуйста, config.settings.prod по умолчанию.

Denis-Shtanskiy replied:
Не в тот ПР закинул)))

ArtemKAF commented on src/backend/api/v1/projects/views.py:
Подтяни в ветку последние изменения из develop. А то этот ПР перезатрет последние фикс, котрый открывает доступ анонимным пользователям к получению информации о сущностях, которые необходимы для фильтрации проектов.

Denis-Shtanskiy replied:
Подтягивал, нужно ещё раз проверить, спасибо

ArtemKAF commented on src/backend/api/v1/profile/views.py:
Из этого файла убери все изменения, чтобы меньше конфликтов было с ПР-ом от Ирины.

Denis-Shtanskiy replied:
Не заметил, сразу другие изменения, спасибо

ArtemKAF commented on src/backend/api/v1/profile/filters.py:
Это нужно либо отрефакторить сразу под те изменения, которые Ирина подготовила в ПР-е.
А лучше сейчас не трогать, и вернуться к этому фильтрсету после того, как будет смерджен ПР от Ирины с переработанными профилями и специалистами.

Denis-Shtanskiy replied:
Я так и написал, что не трогал его

Denis-Shtanskiy committed:
fix: Минимальные исправления
- Поправил profile/views
- Вернул в менеджере прод

Denis-Shtanskiy commented:
В коммите написаны изменения. Ветка девелоп спулена полностью, вохможно придется решать конфликты.

Denis-Shtanskiy committed:
fix: Изменения потерянные из develop
